Reorganize makefile

kravciak opened the pull request:
## Description

- generate target for every testcase from filename
- simplify targets to improve readability

```bash
# create k3d cluster
make cluster

# install cert-manager & kubewarden
make install

# run test by filename or test group
make basic-end-to-end-tests.bats upgrade.bats
make tests

# install fresh kw on new cluster
make reinstall

# remove k3d cluster
make clean
```

viccuad commented on Makefile:
```diff
@@ -56,149 +61,69 @@ bats = RESOURCES_DIR=$(RESOURCES_DIR) \
-	$(delete-cluster)
+# Target all non-destructive tests
+.PHONY: tests
+tests: $(filter-out upgrade.bats, $(TESTS))
```
:heart:

viccuad reviewed (approved):
I find this great, approving!

jvanz commented:
> The changes seems fine to me. But the upgrade tests is failing when I try to run it locally. The command that I'm using to run it is:
> 
> ```
> make cluster install upgrade.bats
> ```
> 
> Furthermore, can you please update the README file with the basic target and how to use them? The file is out of date and we should keep it updated.

My guess is the issue is in this line:

```console
upgrade.bats:: clean cluster
```

After I've aborted the command I could not find any cluster locally. Thus, I think that the cluster is deleted before running the tests. That's make sense if you see the error messages:

```console
pod/kubewarden-controller-5c489689c5-n6vx7 condition met                                                                                                                                                                                     
k3d cluster delete kubewarden-testing                                                                                                                                                                                                        
INFO[0000] Deleting cluster 'kubewarden-testing'                                                                                                                                                                                             
INFO[0002] Deleting 2 attached volumes...                                                                                                                                                                                                    
WARN[0003] Failed to delete volume 'k3d-kubewarden-testing-images' of cluster 'kubewarden-testing': failed to find volume 'k3d-kubewarden-testing-images': Error: No such volume: k3d-kubewarden-testing-images -> Try to delete it manually 
INFO[0003] Removing cluster details from default kubeconfig...                                                                                                                                                                               
INFO[0003] Removing standalone kubeconfig file (if there is one)...                                                                                                                                                                          
INFO[0003] Successfully deleted cluster kubewarden-testing!                                                                                                                                                                                  
./scripts/generate_resources_dir.sh /home/jvanz/suse/kubewarden-end-to-end-tests/resources policies.kubewarden.io/v1 v1                                                                                                                      
RESOURCES_DIR=/home/jvanz/suse/kubewarden-end-to-end-tests/resources/resources_v1 TIMEOUT=5m KUBEWARDEN_CRDS_CHART_OLD_VERSION=1.0.0 KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION=1.1.6 KUBEWARDEN_CONTROLLER_CHART_OLD_VERSION=1.1.0 KUBEWARDEN_CRD
S_CHART_VERSION=1.1.0 KUBEWARDEN_DEFAULTS_CHART_VERSION=1.1.7 KUBEWARDEN_CONTROLLER_CHART_VERSION=1.1.1 KUBEWARDEN_CONTROLLER_CHART_RELEASE=kubewarden-controller KUBEWARDEN_CHARTS_LOCATION=kubewarden CONTROLLER_CHART=kubewarden/kubewarde
n-controller CLUSTER_CONTEXT=k3d-kubewarden-testing  NAMESPACE=kubewarden bats --print-output-on-failure  /home/jvanz/suse/kubewarden-end-to-end-tests/tests/upgrade.bats                                                                    
upgrade.bats                                                                                                                                                                                                                                 
 ✗ [CRD upgrade] Install old Kubewarden                                                                                                                                                                                                      
   (from function `wait_pods' in file tests/common.bash, line 34,                                                                                                                                                                            
    from function `setup' in test file tests/upgrade.bats, line 5)                                                                                                                                                                           
     `wait_pods' failed                                                                                                                                                                                                                      
   /home/jvanz/suse/kubewarden-end-to-end-tests/tests/common.bash: line 3: bats_require_minimum_version: command not found                                                                                                                   
   error: context "k3d-kubewarden-testing" does not exist                                                                                                                                                                                    
   Fail
   error: context "k3d-kubewarden-testing" does not exist
   Fail
   error: context "k3d-kubewarden-testing" does not exist
   Fail
   error: context "k3d-kubewarden-testing" does not exist
   Fail
   error: context "k3d-kubewarden-testing" does not exist
   Fail
   error: context "k3d-kubewarden-testing" does not exist
   Fail
   Godot: pods not running

```

kravciak commented:
>    /home/jvanz/suse/kubewarden-end-to-end-tests/tests/common.bash: line 3: bats_require_minimum_version: command not found                                                                                                                   

What version of bats do you use? You need at least `1.5.0`

For normal tests you run
```bash
make clean   # to remove previous cluster
make cluster # to create k3d cluster
make install # to install kubewarden

# or you can group 3 steps above into
make reinstall

# then run tests
make test1.bats test2.bats
```

Upgrade test is different, it needs to install kubewarden itself because it starts with version -1.
I added dependencies to makefile, so you run just
```bash
# makefile will clean, create cluster, install cert-manager and run upgrade test
make upgrade.bats
```

Currently you need to hack `KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION` to `jq -r ".[0].version"` - because `.[1].version` is broken

I will fix README, thanks for review

kravciak commented:
I updated README and fixed `upgrade.test`. I also removed `prepare` target because it only installed k3d - I don't think it's useful to have target just for that. Can you review again pls?

jvanz commented:
@kravciak for some reason when I run the command `make upgrade.bats` it get stuck. The command never finished. This is my environment:

The last commit:
```
commit 55f511ee0eb3647d403d12bce291f0fb1002316f (HEAD -> martin-main, kravciak/main)
Author: Martin Kravec <kravciak@gmail.com>
Date:   Mon Sep 5 14:45:40 2022 +0200

    Fix Makefile description in README

 README.md | 60 ++++++++++++++++++++++++++++++++++--------------------------
 1 file changed, 34 insertions(+), 26 deletions(-)
```

The change applied in the Makefile:
```
$ git --no-pager diff
diff --git a/Makefile b/Makefile
index f1d1ebe..5f1e43f 100644
--- a/Makefile
+++ b/Makefile
@@ -22,7 +22,7 @@ KUBEWARDEN_HELM_REPO_URL ?= https://charts.kubewarden.io
 KUBEWARDEN_CHARTS_LOCATION ?= kubewarden
 
 KUBEWARDEN_CONTROLLER_CHART_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_CONTROLLER_CHART_RELEASE) --versions -o json --devel | jq -r ".[0].version")
-KUBEWARDEN_CONTROLLER_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_CONTROLLER_CHART_RELEASE) --versions -o json --devel | jq -r ".[1].version")
+KUBEWARDEN_CONTROLLER_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_CONTROLLER_CHART_RELEASE) --versions -o json --devel | jq -r ".[0].version")
 KUBEWARDEN_CONTROLLER_CHART_RELEASE ?= kubewarden-controller
 KUBEWARDEN_CRDS_CHART_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_CRDS_CHART_RELEASE) --versions -o json --devel | jq -r ".[0].version")
 KUBEWARDEN_CRDS_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_CRDS_CHART_RELEASE) --versions -o json --devel | jq -r ".[1].version")****
```

Bats version:
```
$ bats --version
Bats 1.6.0
```

Some errors from the policy server logs:
```
2022-09-05T18:35:22.071026Z ERROR policy_server::worker_pool: cannot spawn worker error="[ghcr.io/kubewarden/policies/capabilities-psp:v0.1.9: settings of policy ghcr.io/kubewarden/policies/capabilities-psp:v0.1.9 are invalid: Some(\ 
"Error invoking settings validation callback: GuestCallFailure(\\\"Guest call failed: Error decoding validation payload {\\\\\\\"allowed_capabilities\\\\\\\":null,\\\\\\\"required_drop_capabilities\\\\\\\":[\\\\\\\"ALL\\\\\\\"],\\\\\ 
\\"default_add_capabilities\\\\\\\":null}: Error(\\\\\\\"invalid type: null, expected a sequence\\\\\\\", line: 1, column: 28)\\\")\")]"                                                                                                  │

2022-09-05T18:35:22.071054Z ERROR policy_server::worker_pool: cannot spawn worker error="[ghcr.io/kubewarden/policies/capabilities-psp:v0.1.9: settings of policy ghcr.io/kubewarden/policies/capabilities-psp:v0.1.9 are invalid: Some(\ 
"Error invoking settings validation callback: GuestCallFailure(\\\"Guest call failed: Error decoding validation payload {\\\\\\\"allowed_capabilities\\\\\\\":null,\\\\\\\"required_drop_capabilities\\\\\\\":[\\\\\\\"ALL\\\\\\\"],\\\\\ 
\\"default_add_capabilities\\\\\\\":null}: Error(\\\\\\\"invalid type: null, expected a sequence\\\\\\\", line: 1, column: 28)\\\")\")]"   
```

kravciak commented:
@jvanz your git diff is wrong - change `KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION`, not `KUBEWARDEN_CONTROLLER_CHART_OLD_VERSION`
```diff
-KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_DEFAULTS_CHART_RELEASE) --versions -o json --devel | jq -r ".[1].version")
+KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_DEFAULTS_CHART_RELEASE) --versions -o json --devel | jq -r ".[0].version")
```

jvanz commented:
> @jvanz your git diff is wrong - change `KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION`, not `KUBEWARDEN_CONTROLLER_CHART_OLD_VERSION`
> 
> ```diff
> -KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_DEFAULTS_CHART_RELEASE) --versions -o json --devel | jq -r ".[1].version")
> +KUBEWARDEN_DEFAULTS_CHART_OLD_VERSION ?= $(shell helm search repo $(KUBEWARDEN_HELM_REPO_NAME)/$(KUBEWARDEN_DEFAULTS_CHART_RELEASE) --versions -o json --devel | jq -r ".[0].version")
> ```

:man_facepalming:
